### ISSUE TITLE
Add an example for NLP (HuggingFace)

### ISSUE THREAD

**Issue by lebrice** [mila-iqia/ResearchTemplate#14]
Potentially interesting:
- https://github.com/NielsRogge/Transformers-Tutorials/blob/master/VisionTransformer/Fine_tuning_the_Vision_Transformer_on_CIFAR_10_with_PyTorch_Lightning.ipynb
- https://lightning.ai/docs/pytorch/LTS/notebooks/lightning_examples/text-transformers.html

## User Research

- This isn't as critical as for the RL example, since HuggingFace is so ubiquitous in NLP that there really isn't a question as to what framework/tool we should be using.
  However, it might be good to check that models / datasets might be best for a template. It's probably safe to assume that wikitext-103 and Llama2/llama3 might be good datasets/models to start with. 

## Design
- [x] (optional, recommended) Design a `TextDataModule` or similar, to encapsulate everything related to the dataset preparation / preprocessing / tokenization
	- This could be a good starting point: https://lightning.ai/docs/pytorch/LTS/notebooks/lightning_examples/text-transformers.html, but ideally we'd like to create a datamodule where the choice of dataset can be changed instead of being hard-coded to GLUE.
	- It might be tricky to keep this algorithm-agnostic, given that the choice of tokenizer is usually tied to the algorithm. Here's one idea though: In the eventual datamodule config, you could have something like `tokenizer: ${instance_attr:algorithm.tokenizer,some_default_tokenizer}`  so that the DataModule will try to use the tokenizer specified by the algorithm, and fallback to the default tokenizer for the dataset.

## Implementation

- [x] (optional, recommended): Write some tests, for both the text datamodule and the lightning module. Some suggestions:
	-  The network forward / backward pass should be reproducible given a seed
	-  Pretrained weights should be loaded at the right location (on $SCRATCH when on a slurm cluster)
	-  Datasets should be loaded a the right location as well
	- Using a pretrained model
	- Overfitting on a batch of data, etc.

- [x] Add support for HuggingFace datasets (`datasets`)
	-  With or without a DataModule, here are some requirements for the implementation: 
	- [ ] Make use of the shared huggingface cache on the Mila cluster (ask @lebrice for more info)
	- [x] Make sure to use $SCRATCH instead of $HOME for the huggingface cache directory.
	- [x] The tokenized datasets should probably be writen to $SCRATCH on first creation, and then copied to $SLURM_TMPDIR before being loaded.

- [x] Add support for HuggingFace models (`transformers`)
	- [ ] (optional) Instead of adding a new algorithm for NLP, start by adapting the current examples so that they can also use transformers as their `network`
		- It might be necessary to extract out the portion of the `training_step` that is specific to classification, and move it into something like a `ClassificationOutputHead` so that the choice of "output head type" is based on the kind of datamodule.  We'd then like to have something like a `NLPOutputHead`
		- There's definitely multiple ways to do the same thing. The "different output heads" idea isn't necessarily the right one. Feel free to discuss / suggest alternatives. 
	- [x] Add a config for an experiment that uses a network and dataset from HF

## Validation
- [x] Run the experiment on the Mila / Drac clusters
	- [ ] (optional) Add a link to a wandb run or report in the PR if possible.
- [x] Make sure that all tests pass (locally, on the self-hosted runner, mila cluster, DRAC clusters)
